server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```
@@ -1382,7 +1424,13 @@ private SearchSourceBuilder parseXContent(
             if (token == XContentParser.Token.FIELD_NAME) {
                 currentFieldName = parser.currentName();
             } else if (token.isValue()) {
-                if (FROM_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                if (PROJECT_ROUTING.match(currentFieldName, parser.getDeprecationHandler()) && searchRequest != null) {
```

I wonder if we need more than this. If searchRequest is null and project_routing is present do we need to throw an error? Or maybe that already happens since then you never all parser.text() on line 1432 if searchRequest is null and then you'll get an error message? But that error message should be clear like "project_routing not supported for this endpoint or an non-serverless environments" and not some general parsing error like "expected END_OBJECT" or something.

So can you also add a REST test to this repo that ensure that project_routing cannot be set for stateful _search and _async_search?

---

The same generic message you get for any invalid input: `Unknown key for a VALUE_STRING in [project_routing]`.